README.md
This is the bison++ part of the project from the now archived 
http://code.google.com/p/flexpp-bisonpp. The original aim of that project was
to fix the original bison++ so that it compiled with gcc 4.5.


ChangeLog
Changes between 1.19-6 and 1.21-8
	- cosmetics...
	- bison.cc, bison.h : generalize MSDOS to WINDOWS and alike.... 
	- bison.cc, bison.h : WINDOWS: simulate alloca with malloc, via macro 
Changes between 1.19-6 and 1.21-7
	- bison.cc, bison.h : use macro to simulate goto in yyparse. for sun c++.
	- bison.cc : mismatch in #if between YY_USE_CLASS and compatibility just
        before including #define tokens. added a #endif and a #ifndef YY_USE_CLASS
	- generate enum to replace enventually the static const
	- output.c, files.c : manage to put #line in generated code, such that the generated code is correctly indicated in the output file. 
Changes between 1.19-5 and 1.21-6
	- bison.h : parameter name forgotten in 
	'virtual void YY_@_ERROR(char *) YY_@_ERROR_BODY;' so inline cannot use
        the parameter. just add the name 'msg' as parameter.
Changes between 1.19-3 and 1.21-5
	- send on internet as patch 5
	- patch according to release 1.21 of bison :
Mon Sep  6 15:32:32 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * Version 1.22 released.

        * mkinstalldirs: New file.

        * Makefile.in (dist): Use .gz for extension, not .z.
        (DISTFILES): New variable.
        (dist): Use it instead of explicit file list.
        Try to link each file separately, then copy file if ln fails.
        (installdirs): Use mkinstalldirs script.

Thu Jul 29 20:35:02 1993  David J. MacKenzie  (djm@wookumz.gnu.ai.mit.edu)

	* Makefile.in (config.status): Run config.status --recheck, not
	configure, to get the right args passed.

Sat Jul 24 04:00:52 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple (yyparse): Init yychar1 to avoid warning.

Sun Jul  4 16:05:58 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple (yyparse): Don't set yyval when yylen is 0.

Sat Jun 26 15:54:04 1993  David J. MacKenzie  (djm@wookumz.gnu.ai.mit.edu)

	* getargs.c (getargs): Exit after printing the version number.
	Add --help and -h options.
	(usage): New function.

Fri Jun 25 15:11:25 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* getargs.c (longopts): Allow `output' as an alternative.

Wed Jun 16 17:02:37 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple (yyparse): Conditionalize the entire call to yyoverflow,
	not just two arguments in it.

Thu Jun  3 13:07:19 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple [__hpux] (alloca): Don't specify arg types.

Fri May  7 05:53:17 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * Makefile.in (install): Depend on `uninstall' and `installdirs'.
	(installdirs): New target.

Wed Apr 28 15:15:15 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * reader.c:  Remove declaration of atoi.

Fri Apr 23 12:29:20 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * new.h [!__STDC__] (FREE): Check x != 0.
        Make expr to call `free' evaluate to 0.

Tue Apr 20 01:43:44 1993  David J. MacKenzie  (djm@kropotkin.gnu.ai.mit.edu)

	* files.c [MSDOS]: Use xmalloc, not malloc.
	* allocate.c (xmalloc): Renamed from mallocate.  Remove old wrapper.
	* conflicts.c, symtab.c, files.c, LR0.c, new.h: Change callers.
	* allocate.c (xrealloc): New function.
	* new.h: Declare it.
	* lex.c, reader.c: Use it.

Sun Apr 18 00:45:56 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * Version 1.21 released.

        * reader.c : Don't declare `realloc'.

        * Makefile.in (bison.s1): use `rm -f' since it's quieter.
        (dist): make gzipped tar file.

Fri Apr 16 21:24:10 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * Makefile.in (Makefile, config.status, configure): New targets.

Thu Apr 15 15:37:28 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* main.c: Don't declare `abort'.

	* files.c: Don't declare `exit'.

Thu Apr 15 02:42:38 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * configure.in: Add AC_CONST.

Wed Apr 14 00:51:17 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* Makefile.in (all): Depend on bison.s1.

Tue Apr 13 14:52:32 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* Version 1.20 released.

Wed Mar 24 21:45:47 1993  Richard Stallman  (rms@wookumz.gnu.ai.mit.edu)

	* output.c (output_headers): Rename yynerrs if -p.

Thu Mar 18 00:02:17 1993  Noah Friedman  (friedman@nutrimat.gnu.ai.mit.edu)

        * system.h: Don't try to include stdlib.h unless HAVE_STDLIB_H is
          defined.

        * configure.in: Check for stdlib.h.

Wed Mar 17 14:44:27 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple [__hpux, not __GNUC__]: Declare alloca.
	(yyparse): When printing the expected token types for an error,
	Avoid negative indexes in yycheck and yytname.

Sat Mar 13 23:31:25 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* Makefile.in (files.o, .c.o): Put CPPFLAGS and CFLAGS last.

Mon Mar  1 17:49:08 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* bison.simple: Test __sgi like __sparc.

Wed Feb 17 00:04:13 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* conflicts.c (resolve_sr_conflict): Add extra parens in alloca call.

	* bison.simple [__GNUC__] (yyparse): Declare with prototype.

Fri Jan 15 13:15:17 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* conflicts.c (print_reduction): Near end, increment fp2 when mask
	recycles.

Wed Jan 13 04:15:03 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* Makefile.in (bison.s1): New target.  Modifies bison.simple.
	(install): Install bison.s1, without changing it.
	(clean): Delete bison.s1.

Mon Jan  4 20:35:58 1993  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* reader.c (reader): Put Bison version in comment in output file.

Tue Dec 22 19:00:58 1992  Richard Stallman  (rms@mole.gnu.ai.mit.edu)

	* files.c (openfiles): Use .output, not .out, for outfile,
	regardless of spec_name_prefix.

Tue Dec 15 19:22:11 1992  Richard Stallman  (rms@mole.gnu.ai.mit.edu)


lex.c
/* Token-reader for Bison's input parser,
   Copyright (C) 1984, 1986, 1989 Free Software Foundation, Inc.

This file is part of Bison, the GNU Compiler Compiler.

Bison is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

Bison is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Bison; see the file COPYING.  If not, write to
the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.  */


/* 
   lex() is the entry point.  It is called from reader.c.
   It returns one of the token-type codes defined in lex.h.
   When an identifier is seen, the code IDENTIFIER is returned
   and the name is looked up in the symbol table using symtab.c;
   symval is set to a pointer to the entry found.  */

#include <stdio.h>
#include <ctype.h>
#include "system.h"
#include "files.h"
#include "symtab.h"
#include "lex.h"
#include "new.h"


extern int lineno;
extern int translations;

int parse_percent_token();

extern void fatals();
extern void fatal();

/* Buffer for storing the current token.  */
char *token_buffer;

/* Allocated size of token_buffer, not including space for terminator.  */
static int maxtoken;

bucket *symval;
int numval;

static int unlexed;		/* these two describe a token to be reread */
static bucket *unlexed_symval;	/* by the next call to lex */


void
init_lex()
{
  maxtoken = 100;
  token_buffer = NEW2 (maxtoken + 1, char);
  unlexed = -1;
}


static char *
grow_token_buffer (p)
     char *p;
{
  int offset = p - token_buffer;
  maxtoken *= 2;
  token_buffer = (char *) xrealloc(token_buffer, maxtoken + 1);
  return token_buffer + offset;
}


int
skip_white_space()
{
  register int c;
  register int inside;

  c = getc(finput);

  for (;;)
    {
      int cplus_comment;

      switch (c)
	{
	case '/':
	  c = getc(finput);
	  if (c != '*' && c != '/')
	    fatals("unexpected `/%c' found",c);
	  cplus_comment = (c == '/');

	  c = getc(finput);

	  inside = 1;
	  while (inside)
	    {
	      if (!cplus_comment && c == '*')
		{
		  while (c == '*')
		    c = getc(finput);

		  if (c == '/')
		    {
		      inside = 0;
		      c = getc(finput);
		    }
		}
	      else if (c == '\n')
		{
		  lineno++;
		  if (cplus_comment)
		    inside = 0;
		  c = getc(finput);
		}
	      else if (c == EOF)
		fatal("unterminated comment");
	      else
		c = getc(finput);
	    }

	  break;

	case '\n':
	  lineno++;

	case ' ':
	case '\t':
	case '\f':
	  c = getc(finput);
	  break;

	default:
	  return (c);
	}
    }
}


void
unlex(token)
int token;
{
  unlexed = token;
  unlexed_symval = symval;
}



int
lex()
{
  register int c;
  register char *p;

  if (unlexed >= 0)
    {
      symval = unlexed_symval;
      c = unlexed;
      unlexed = -1;
      return (c);
    }

  c = skip_white_space();

  switch (c)
    {
    case EOF:
      return (ENDFILE);

    case 'A':  case 'B':  case 'C':  case 'D':  case 'E':
    case 'F':  case 'G':  case 'H':  case 'I':  case 'J':
    case 'K':  case 'L':  case 'M':  case 'N':  case 'O':
    case 'P':  case 'Q':  case 'R':  case 'S':  case 'T':
    case 'U':  case 'V':  case 'W':  case 'X':  case 'Y':
    case 'Z':
    case 'a':  case 'b':  case 'c':  case 'd':  case 'e':
    case 'f':  case 'g':  case 'h':  case 'i':  case 'j':
    case 'k':  case 'l':  case 'm':  case 'n':  case 'o':
    case 'p':  case 'q':  case 'r':  case 's':  case 't':
    case 'u':  case 'v':  case 'w':  case 'x':  case 'y':
    case 'z':
    case '.':  case '_':
      p = token_buffer;
      while (isalnum(c) || c == '_' || c == '.')
	{
	  if (p == token_buffer + maxtoken)
	    p = grow_token_buffer(p);

	  *p++ = c;
	  c = getc(finput);
	}

      *p = 0;
      ungetc(c, finput);
      symval = getsym(token_buffer);
      return (IDENTIFIER);

    case '0':  case '1':  case '2':  case '3':  case '4':
    case '5':  case '6':  case '7':  case '8':  case '9':
      {
	numval = 0;

	while (isdigit(c))
	  {
	    numval = numval*10 + c - '0';
	    c = getc(finput);
	  }
	ungetc(c, finput);
	return (NUMBER);
      }

    case '\'':
      translations = -1;

      /* parse the literal token and compute character code in  code  */

      c = getc(finput);
      {
	register int code = 0;

	if (c == '\\')
	  {
	    c = getc(finput);

	    if (c <= '7' && c >= '0')
	      {
		while (c <= '7' && c >= '0')
		  {
		    code = (code * 8) + (c - '0');
		    c = getc(finput);
		    if (code >= 256 || code < 0)
		      fatals("malformatted literal token `\\%03o'", code);
		  }
	      }
	    else
	      {
		if (c == 't')
		  code = '\t';
		else if (c == 'n')
		  code = '\n';
		else if (c == 'a')
		  code = '\007';
		else if (c == 'r')
		  code = '\r';
		else if (c == 'f')
		  code = '\f';
		else if (c == 'b')
		  code = '\b';
		else if (c == 'v')
		  code = 013;
		else if (c == 'x')
		  {
		    c = getc(finput);
		    while ((c <= '9' && c >= '0')
			   || (c >= 'a' && c <= 'z')
			   || (c >= 'A' && c <= 'Z'))
		      {
			code *= 16;
			if (c <= '9' && c >= '0')
			  code += c - '0';
			else if (c >= 'a' && c <= 'z')
			  code += c - 'a' + 10;
			else if (c >= 'A' && c <= 'Z')
			  code += c - 'A' + 10;
			if (code >= 256 || code<0)/* JF this said if(c>=128) */
			  fatals("malformatted literal token `\\x%x'",code);
			c = getc(finput);
		      }
		    ungetc(c, finput);
		  }
		else if (c == '\\')
		  code = '\\';
		else if (c == '\'')
		  code = '\'';
		else if (c == '\"')	/* JF this is a good idea */
		  code = '\"';
		else
		  {
		    if (c >= 040 && c <= 0177)
		      fatals ("unknown escape sequence `\\%c'", c);
		    else
		      fatals ("unknown escape sequence: `\\' followed by char code 0x%x", c);
		  }

		c = getc(finput);
	      }
	  }
	else
	  {
	    code = c;
	    c = getc(finput);
	  }
	if (c != '\'')
	  fatal("multicharacter literal tokens not supported");

	/* now fill token_buffer with the canonical name for this character
	   as a literal token.  Do not use what the user typed,
	   so that '\012' and '\n' can be interchangeable.  */

	p = token_buffer;
	*p++ = '\'';
	if (code == '\\')
	  {
	    *p++ = '\\';
	    *p++ = '\\';
	  }
	else if (code == '\'')
	  {
	    *p++ = '\\';
	    *p++ = '\'';
	  }
	else if (code >= 040 && code != 0177)
	  *p++ = code;
	else if (code == '\t')
	  {
	    *p++ = '\\';
	    *p++ = 't';
	  }
	else if (code == '\n')
	  {
	    *p++ = '\\';
	    *p++ = 'n';
	  }
	else if (code == '\r')
	  {
	    *p++ = '\\';
	    *p++ = 'r';
	  }
	else if (code == '\v')
	  {
	    *p++ = '\\';
	    *p++ = 'v';
	  }
	else if (code == '\b')
	  {
	    *p++ = '\\';
	    *p++ = 'b';
	  }
	else if (code == '\f')
	  {
	    *p++ = '\\';
	    *p++ = 'f';
	  }
        else
	  {
	    *p++ = code / 0100 + '0';
	    *p++ = ((code / 010) & 07) + '0';
	    *p++ = (code & 07) + '0';
	  }
	*p++ = '\'';
	*p = 0;
	symval = getsym(token_buffer);
	symval->class = STOKEN;
	if (! symval->user_token_number)
	  symval->user_token_number = code;
	return (IDENTIFIER);
      }

    case ',':
      return (COMMA);

    case ':':
      return (COLON);

    case ';':
      return (SEMICOLON);

    case '|':
      return (BAR);

    case '{':
      return (LEFT_CURLY);

    case '=':
      do
	{
	  c = getc(finput);
	  if (c == '\n') lineno++;
	}
      while(c==' ' || c=='\n' || c=='\t');

      if (c == '{')
      	return(LEFT_CURLY);
      else
	{
	  ungetc(c, finput);
	  return(ILLEGAL);
	}

    case '<':
      p = token_buffer;
      c = getc(finput);
      while (c != '>')
	{
	  if (c == '\n' || c == EOF)
	    fatal("unterminated type name");

	  if (p == token_buffer + maxtoken)
	    p = grow_token_buffer(p);

	  *p++ = c;
	  c = getc(finput);
	}
      *p = 0;
      return (TYPENAME);
	    

    case '%':
      return (parse_percent_token());

    default:
      return (ILLEGAL);
    }
}


/* parse a token which starts with %.  Assumes the % has already been read and discarded.  */

int
parse_percent_token ()
{
  register int c;
  register char *p;

  p = token_buffer;
  c = getc(finput);

  switch (c)
    {
    case '%':
      return (TWO_PERCENTS);

    case '{':
      return (PERCENT_LEFT_CURLY);

    case '<':
      return (LEFT);

    case '>':
      return (RIGHT);

    case '2':
      return (NONASSOC);

    case '0':
      return (TOKEN);

    case '=':
      return (PREC);
    }
  if (!isalpha(c))
    return (ILLEGAL);

  while (isalpha(c) || c == '_')
    {
      if (p == token_buffer + maxtoken)
	p = grow_token_buffer(p);

      *p++ = c;
      c = getc(finput);
    }

  ungetc(c, finput);

  *p = 0;

  if (strcmp(token_buffer, "token") == 0
      ||
      strcmp(token_buffer, "term") == 0)
    return (TOKEN);
  else if (strcmp(token_buffer, "nterm") == 0)
    return (NTERM);
  else if (strcmp(token_buffer, "type") == 0)
    return (TYPE);
  else if (strcmp(token_buffer, "guard") == 0)
    return (GUARD);
  else if (strcmp(token_buffer, "union") == 0)
    return (UNION);
  else if (strcmp(token_buffer, "expect") == 0)
    return (EXPECT);
  else if (strcmp(token_buffer, "start") == 0)
    return (START);
  else if (strcmp(token_buffer, "left") == 0)
    return (LEFT);
  else if (strcmp(token_buffer, "right") == 0)
    return (RIGHT);
  else if (strcmp(token_buffer, "nonassoc") == 0
	   ||
	   strcmp(token_buffer, "binary") == 0)
    return (NONASSOC);
  else if (strcmp(token_buffer, "semantic_parser") == 0)
    return (SEMANTIC_PARSER);
  else if (strcmp(token_buffer, "pure_parser") == 0)
    return (PURE_PARSER);
  else if (strcmp(token_buffer, "prec") == 0)
    return (PREC);
  else if (strcmp(token_buffer, "name") == 0)
    return (PARSER_NAME);
  else if (strcmp(token_buffer, "define") == 0)
    return (DEFINE_SYM);
  else if (strcmp(token_buffer, "header") == 0)
    { c=getc(finput);
      if(c=='{')
          return (PERCENT_LEFT_CURLY_HEADER);
      else return (ILLEGAL);
    }
  else return (ILLEGAL);
}
